docs: add CONTRIBUTING.md file for docs (#758)
keptn/lifecycle-toolkit: docs/CONTRIBUTING.md

to check your formatting, you have the following options:

   - If you are using an IDE to author the markdown text,
     use the markdown preview browser for the IDE.
   - You can push your changes to GitHub
     and use the GitHub previewer (*View Page*).
   - You can install and use the
     [grip](https://github.com/joeyespo/grip/blob/master/README.md) previewer
     to view the rendered content locally.
     When *grip* is installed,
     you can format the specified file locally
     by running the following in its own shell:

     ```
     grip <file>.md
     ```
     Point your browser at `localhost:6419` to view the formatted file.
     The document updates automatically
     each time you write your changes to disk.

## Interacting with github

The documentation source is stored on github.com
and you use the standard github facilities to modify it.
The following notes are provided for people who need a little help.

### Fork and clone the repository

Perform the following steps to create a copy of this repository on your local machine:

1. Fork the Keptn repository:

     - Log into GitHub (or create a GitHub account and then log into it).

>      - Go to the [Keptn docs repository](https://github.com/keptn/keptn.github.io).

Reviewer: ```suggestion
     - Go to the [Keptn lifecycle-toolkit repository](https://github.com/keptn/lifecycle-toolkit).
```

Author: done

     - Click the **Fork** button at the top of the screen.
     - Choose the user for the fork from the options you are given,
       usually your GitHub ID.

   A copy of this repository is now available in your GitHub account.

2. Get the string to use when cloning your fork:

     - Click the green "Code" button on the UI page.
     - Select the protocol to use for this clone (either HTTPS or SSH).
     - A box is displayed that gives the URL for the selected protocol.
       Click the icon at the right end of that box to copy that URL.

3.  Run the **git clone** command from the shell of a local directory
    to clone the forked repository to a directory on your local machine,
    pasting in the URl you saved in the previous step:
    ```

>     git clone https://github.com/<UserName>/keptn.github.io

Reviewer: ```suggestion
    git clone https://github.com/<UserName>/lifecycle-toolkit
```

    ```
    or
    ```
    git clone git@github.com:<UserName>/keptn.github.io.git
    ```
    Where <*UserName*> is your GitHub username.
    The keptn.github.io directory is now available in the local directory.

4. Associate your clone with `upstream`.
   To do this, use the same string you used to clone your fork.

   - Be sure that you are in the root folder of the project
     and run *git status* to confirm that you are on the `main` branch.
   - Type the following to associate `upstream` with your clone,
     pasting in the string for the main repo that you copied above.:
     ```

>      git remote add upstream https://github.com/keptn/keptn.github.io.git 

Reviewer: ```suggestion
     git remote add upstream https://github.com/keptn/lifecycle-toolkit.git 
```

Author: done

### Create a new branch and make your changes

Now you have your upstream setup in your local machine.
You need to create a local branch where you will make your changes.
Be sure that your branch is based on and sync'ed with `main`,
unless you intend to create a derivative PR:

The following sequence of commands does that:

```
git checkout main
git pull upstream main
git push origin main
git checkout -b <my-new-branch>
```

Now you can make your changes, build them locally to check formatting,
then create a PR to add these changes to the documentation set

### Submitting new content through a pull request

If you have forked and cloned the repository,
you can modify the documentation or add new material
by editing the markdown file(s) in the local clone of your fork
and then submitting a _pull request (PR)_.

Note that you can also modify the source using the GitHub editor.
This is very useful when you want to fix a typo or make a small editorial change
but, if you are doing significant writing,
it is generally better to work on files in your local clone.

3. Execute the following and check the output to ensure that your branch is set up correctly:
   ```
   git status
   ```

> 3. Do the writing you want to do in your local branch, checking the formatted version at `localhost:1313`.

Reviewer: ```suggestion
3. Do the writing you want to do in your local branch, checking the formatted version at `localhost:1314`.
```

Author: It's actually 1314/docs-dev.  Fixed

4. Add and commit your changes.  Here, we commit all modified files but you can specify individual files to the `git add` command.
The `git commit -s` command commits the files and signs that you are contributing this intellectual property to the Keptn project.
   ```
   git add .
   git commit -s
   ```

   Use vi commands to add a description of the PR (should be 80 characters or less) to the commit. This will be displayed as the title of the PR in listings.
You can add multiple lines explaining the PR here but, in general, it is better to only supply the PR title here; you can add more information and edit the PR title when you create the PR on the GitHub UI page.

5. Push your branch to github:
   - If you cloned your fork to use SSH, the command is:
     ```
     git push --set-upstream origin <branch-name>
     ```

     Note that you can just issue the `git push` command.
Git responds with an error message that gives you the full command line to use; you can copy this command and paste it into your shell to push the content.
   - If you cloned your fork to use HTTP, the command is:
     ```
     git push <need options/arguments>
     ```

> 6. Create the PR by going to the [keptn.github.io](https://github.com/keptn/keptn.github.io) GitHub repository.

Reviewer: ```suggestion
6. Create the PR by going to the [lifecycle-toolkit](https://github.com/keptn/lifecycle-toolkit) GitHub repository.
```

Author: done

   * If a reviewer makes a GitHub suggestion and you agree with the change, just click "Accept this change" to create a commit for that modification.  You can also group several suggestions into a single commit using the GitHub tools.
   * You can make other changes using the GitHub editor or you can work in your local branch to make modifications.

      * If changes have been made using the GitHub editor, you will need to do a `git pull` request to pull those commits back to your local branch before you push the new changes.
      * After modifying the local source, issue the `git add .`, `git commit`, and `git push` commands to push your changes to github.

8. When your PR has the appropriate approvals, it will be merged and the revised content should be published on the web site within a few minutes.

9. When your PR has been approved and merged,
   you can delete your local branch with the following command:
   ```
   git branch -d <branch-name>
   ```

### Developer Certification of Origin (DCO)

Licensing is very important to open source projects. It helps ensure the software continues to be available under the terms that the author desired.

> Keptn uses [Apache License 2.0](https://github.com/keptn/keptn.github.io/blob/master/LICENSE) to strike a balance between open contributions and allowing you to use the software however you would like to.

Reviewer: ```suggestion
Keptn uses [Apache License 2.0](https://github.com/keptn/lifecycle-toolkit/blob/main/LICENSE) to strike a balance between open contributions and allowing you to use the software however you would like to.
```

Author: done

    this project or the open source license(s) involved.
```

#### DCO Sign-Off Methods

The DCO requires a sign-off message in the following format to appear on each commit in the pull request:

```
Signed-off-by: Humpty Dumpty <humpty.dumpty@example.com>
```

The DCO text can either be manually added to your commit body, or you can add either **-s** or **--signoff** to your usual git commit commands. If you are using the GitHub UI to make a change you can add the sign-off message directly to the commit message when creating the pull request. If you forget to add the sign-off you can also amend a previous commit with the sign-off by running **git commit --amend -s**. If you've pushed your changes to GitHub already you'll need to force push your branch after this with **git push -f**.

**ATTRIBUTION**:

- https://probot.github.io/apps/dco/
- https://github.com/opensearch-project/common-utils/blob/main/CONTRIBUTING.md
- https://code.asam.net/simulation/wiki/-/wikis/docs/project_guidelines/ASAM-DCO?version_id=c510bffb1195dc04deb9db9451112669073f0ba5
- https://thesofproject.github.io/latest/contribute/contribute_guidelines.html

## Source file structure [NOT YET UPDATED FOR KLT!]

> The source files for the [Keptn Documentation](https://keptn.sh/docs/) are stored under the *content/docs* directory in the repository.

Reviewer: ```suggestion
The source files for the [Keptn Lifecycle Toolkit Documentation](https://lifecycle.keptn.sh/docs) are stored under the *docs/content/en/docs* directory in the repository.
```

> The build strategy is to build everything except for the files/folders that are listed in the `ignoreFiles` array in the [config.toml](https://github.com/keptn/keptn.github.io/blob/master/config.toml) file.

Reviewer: The configuration for LT is written in yaml and not in toml

Author: I don't find either the yaml or toml file in the repo right now.  @aepfli can you clarify?

Reviewer 2: I can't clarify a lot here. I can only state the facts: keptn.github.io uses `toml`, and klt-docs uses `yaml` for the site configuration. We want to try to unify this somewhat in the future to reduce the differences within the project. But currently, I don't see a focus on that.

Author: Also mentioned the hidden: true string in the metadata here.

# Contribute to the Keptn documentation

This document provides information about contributing to
the [Keptn Lifecycle Toolkit documentation](https://lifecycle.keptn.sh/docs/),
which is part of the [Keptn](https://keptn.sh) website.

The Keptn Lifecycle Toolkit documentation is authored with
[markdown](https://www.markdownguide.org/basic-syntax/)
and rendered using the Hugo
[Docsy](https://www.docsy.dev/) theme.

We welcome and encourage contributions of all levels.
You can make modifications using the GitHub editor;
this works well for small modifications but,
if you are making significant changes,
you may find it better to fork and clone the repository
and make changes using the text editor or IDE of your choice.
You can also run the Docsy based website locally
to check the rendered documentation
and then push your changes to the repository as a pull request.

If you need help getting started,
feel free to ask for help on the `#help-contributing` or `#keptn-docs` channels on the [Keptn Slack](https://keptn.sh/community/#slack).
We were all new to this once and are happy to help you!

## Guidelines for contributing

> * Always fork the repository then clone that fork to your local system

Reviewer: I think we should also mention that PRs in the repo follow the semantic commit scheme. Therefore doc PRs should start with "docs:"

Author: Good idea!  I added it to the information about creating the PR, somewhere around line 241.

## Building the documentation locally

You can run Docsy locally so that you can view the formatted version
of what you are writing before you push it to github.
We provide a Docsy run environment in a Docker container,
which simplifies the set up
and makes it easier to upgrade your local build environment
as the software is updated.

To set up a local Docsy build:

1. Install Docker Desktop:

   * [Install on macOS](https://docs.docker.com/desktop/install/mac-install/)
   * [Install on Linux](https://docs.docker.com/desktop/install/linux-install/)
   * [Install on Windows](https://docs.docker.com/desktop/install/windows-install/)

1. Clone the Keptn Docsy repo:
   ```
   make clone
   ```
   **Note:** To utilize the `makefile`, you must have GNU **make**
   available on your local machine.
   Versions are available for all the usual Operating Systems.

1. Execute the following command from the `docs` folder of your clone:
   ```
   make server
   ```
   It will continue running in its own shell.

1. Start contributing!
Note that Hugo updates the rendered documentation each time you write the file.

1. Enter the following in a browser to view the website:
    ```

>     http://localhost:1313/docs-dev/

Reviewer: ```suggestion
    http://localhost:1314/docs-dev/
```

Author: done

### Fork and clone the repository

Perform the following steps to create a copy of this repository on your local machine:

1. Fork the Keptn repository:

     - Log into GitHub (or create a GitHub account and then log into it).
     - Go to the [Keptn lifecycle-toolkit repository](https://github.com/keptn/lifecycle-toolkit).
     - Click the **Fork** button at the top of the screen.
     - Choose the user for the fork from the options you are given,
       usually your GitHub ID.

   A copy of this repository is now available in your GitHub account.

2. Get the string to use when cloning your fork:

     - Click the green "Code" button on the UI page.
     - Select the protocol to use for this clone (either HTTPS or SSH).
     - A box is displayed that gives the URL for the selected protocol.
       Click the icon at the right end of that box to copy that URL.

3.  Run the **git clone** command from the shell of a local directory
    to clone the forked repository to a directory on your local machine,
    pasting in the URl you saved in the previous step:
    ```
    git clone https://github.com/<UserName>/lifecycle-toolkit
    ```
    or
    ```

>     git clone git@github.com:<UserName>/keptn.github.io.git

Reviewer: ```suggestion
    git clone git@github.com:<UserName>/lifecycle-toolkit.git
```

    ```
    Where <*UserName*> is your GitHub username.

>     The keptn.github.io directory is now available in the local directory.

Reviewer: ```suggestion
    The lifecycle-toolkit directory is now available in the local directory.
```

Author: done

```
Signed-off-by: Humpty Dumpty <humpty.dumpty@example.com>
```

The DCO text can either be manually added to your commit body, or you can add either **-s** or **--signoff** to your usual git commit commands. If you are using the GitHub UI to make a change you can add the sign-off message directly to the commit message when creating the pull request. If you forget to add the sign-off you can also amend a previous commit with the sign-off by running **git commit --amend -s**. If you've pushed your changes to GitHub already you'll need to force push your branch after this with **git push -f**.

**ATTRIBUTION**:

- https://probot.github.io/apps/dco/
- https://github.com/opensearch-project/common-utils/blob/main/CONTRIBUTING.md
- https://code.asam.net/simulation/wiki/-/wikis/docs/project_guidelines/ASAM-DCO?version_id=c510bffb1195dc04deb9db9451112669073f0ba5
- https://thesofproject.github.io/latest/contribute/contribute_guidelines.html

## Source file structure [Preliminary]

The source files for the [Keptn Lifecycle Toolkit](https://lifecycle.keptn.sh/docs) are stored under the *docs/content/en/docs* directory in the repository.
The build strategy is to build everything except for:

> * The files/folders that are listed in the `ignoreFiles` array in the [config.toml](https://github.com/keptn/keptn.github.io/blob/master/config.toml) file

Reviewer: I think this is not necessary for the lifecycle-toolkit docs as the versions are appended automatically and the config file only exists in the target repo

Author: So where is the list of files for git to ignore -- like .*.sw?
